ogre2/src/Ogre2COMVisual.cc

```diff
@@ -0,0 +1,218 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ignition/rendering/ogre2/Ogre2COMVisual.hh"
+#include "ignition/rendering/ogre2/Ogre2Material.hh"
+#include "ignition/rendering/ogre2/Ogre2DynamicRenderable.hh"
+
+#ifdef _MSC_VER
+  #pragma warning(push, 0)
+#endif
+#include <OgreSceneNode.h>
+#ifdef _MSC_VER
+  #pragma warning(pop)
+#endif
+
+class ignition::rendering::Ogre2COMVisualPrivate
+{
+  /// \brief Grid materal
+  public: Ogre2MaterialPtr material = nullptr;
+
+  /// \brief Lines that make the cross marking the center of mass.
+  public: std::shared_ptr<Ogre2DynamicRenderable> crossLines = nullptr;
+
+  /// \brief Sphere visual marking the center of mass
+  public: VisualPtr sphereVis = nullptr;
+};
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::Ogre2COMVisual()
+  : dataPtr(new Ogre2COMVisualPrivate)
+{
+}
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::~Ogre2COMVisual()
+{
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::PreRender()
+{
+  if (this->dirtyCOMVisual)
+  {
+    this->CreateVisual();
+    this->dirtyCOMVisual = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Init()
+{
+  BaseCOMVisual::Init();
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Destroy()
+{
+  if (this->dataPtr->sphereVis != nullptr)
+  {
+    this->dataPtr->sphereVis->Destroy();
+    this->dataPtr->sphereVis.reset();
+  }
+
+  if (this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines->Destroy();
+    this->dataPtr->crossLines.reset();
+  }
+
+  if (this->dataPtr->material && this->Scene())
+  {
+    this->Scene()->DestroyMaterial(this->dataPtr->material);
+    this->dataPtr->material.reset();
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::CreateVisual()
+{
+  if (!this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines.reset(
+        new Ogre2DynamicRenderable(this->Scene()));
+    this->ogreNode->attachObject(this->dataPtr->crossLines->OgreObject());
+  }
+
+  if (!this->dataPtr->sphereVis)
+  {
+    this->dataPtr->sphereVis = this->Scene()->CreateVisual();
+    this->dataPtr->sphereVis->AddGeometry(this->Scene()->CreateSphere());
+    this->dataPtr->sphereVis->SetMaterial("Default/CoM");
+    this->dataPtr->sphereVis->SetInheritScale(false);
+    this->AddChild(this->dataPtr->sphereVis);
+  }
+
+  // Compute radius of sphere with density of lead and equivalent mass.
+  double sphereRadius;
+  double dLead = 11340;
```

Comment:
what's this number ?

Reply:
It's the density of Lead that is used to scale the sphere with equivalent mass